File: examples/runners/python/runner.py
#!/usr/bin/env python3
"""
Minimal external-user Python SDK runner for WireParity.

WireParity stdin/stdout IPC protocol:
1. Reads a single JSON line from stdin:
   { "operationId": "...", "inputs": { ... }, "targetUrl": "..." }
2. Translates WireParity IR values to native values and executes the HTTP call.
3. Writes a single JSON line to stdout:
   { "success": true } or { "success": false, "error": "..." }
"""

import json
import sys
import urllib.parse
import urllib.request


def ir_value_to_py(val):
    if not isinstance(val, dict):
        return val
    kind = val.get("kind")
    if kind in ("string", "integer", "number", "boolean", "date", "date-time", "enum"):
        return val.get("value")
    if kind == "null":
        return None
    if kind == "array":
        return [ir_value_to_py(item) for item in val.get("items", [])]
    if kind == "object":
        return {k: ir_value_to_py(v) for k, v in val.get("fields", {}).items()}
    return val.get("value")


def execute_operation(req):
    operation_id = req["operationId"]
    inputs = req.get("inputs", {})
    target_url = req["targetUrl"].rstrip("/")

    if operation_id == "listPets":
        query_params = inputs.get("queryParams", {})
        limit = ir_value_to_py(query_params.get("limit")) if "limit" in query_params else None
        url = f"{target_url}/pets"
        if limit is not None:
            url += f"?limit={urllib.parse.quote(str(limit))}"
        request = urllib.request.Request(url, method="GET")
        with urllib.request.urlopen(request):
            pass

    elif operation_id == "createPet":
        body = ir_value_to_py(inputs.get("body", {}))
        data = json.dumps(body).encode("utf-8")
        request = urllib.request.Request(
            f"{target_url}/pets",
            data=data,
            headers={"Content-Type": "application/json"},
            method="POST",
        )
        with urllib.request.urlopen(request):
            pass

    elif operation_id == "getPetById":
        path_params = inputs.get("pathParams", {})
        pet_id = ir_value_to_py(path_params.get("petId", ""))
        quoted_id = urllib.parse.quote(str(pet_id), safe="")
        request = urllib.request.Request(f"{target_url}/pets/{quoted_id}", method="GET")
        with urllib.request.urlopen(request):
            pass

    else:
        raise ValueError(f"Unsupported operation: {operation_id}")


def main():
    line = sys.stdin.read().strip()
    if not line:
        sys.stdout.write(json.dumps({"success": False, "error": "Empty input"}) + "\n")
        return

    try:
        req = json.loads(line)
        execute_operation(req)
        sys.stdout.write(json.dumps({"success": True}) + "\n")
    except Exception as e:
        sys.stdout.write(json.dumps({"success": False, "error": str(e)}) + "\n")


if __name__ == "__main__":
    main()

File: examples/runners/typescript/runner.ts
#!/usr/bin/env node
/**
 * Minimal external-user TypeScript SDK runner for WireParity.
 *
 * WireParity stdin/stdout IPC protocol:
 * 1. Reads a single JSON line from stdin:
 *    { "operationId": "...", "inputs": { ... }, "targetUrl": "..." }
 * 2. Translates WireParity IR values to native values and executes the HTTP call.
 * 3. Writes a single JSON line to stdout:
 *    { "success": true } or { "success": false, "error": "..." }
 */

// Helper to convert WireParity IR values to native JavaScript values
function irValueToJs(val: any): any {
  if (!val || typeof val !== "object") return val;
  if ("value" in val) return val.value;
  if (val.kind === "null") return null;
  if (val.kind === "array") return (val.items || []).map(irValueToJs);
  if (val.kind === "object") {
    const obj: Record<string, any> = {};
    for (const [k, v] of Object.entries(val.fields || {})) {
      obj[k] = irValueToJs(v);
    }
    return obj;
  }
  return val;
}

async function executeOperation(req: {
  operationId: string;
  inputs: any;
  targetUrl: string;
}): Promise<void> {
  const { operationId, inputs, targetUrl } = req;
  const baseUrl = targetUrl.replace(/\/+$/, "");

  switch (operationId) {
    case "listPets": {
      const limit = inputs.queryParams?.limit
        ? irValueToJs(inputs.queryParams.limit)
        : undefined;
      const url = new URL(`${baseUrl}/pets`);
      if (limit !== undefined) {
        url.searchParams.set("limit", String(limit));
      }
      await fetch(url.toString(), { method: "GET" });
      break;
    }

    case "createPet": {
      const body = inputs.body ? irValueToJs(inputs.body) : {};
      await fetch(`${baseUrl}/pets`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(body),
      });
      break;
    }

    case "getPetById": {
      const petId = inputs.pathParams?.petId
        ? irValueToJs(inputs.pathParams.petId)
        : "";
      await fetch(`${baseUrl}/pets/${encodeURIComponent(String(petId))}`, {
        method: "GET",
      });
      break;
    }

    default:
      throw new Error(`Unsupported operation: ${operationId}`);
  }
}

async function main(): Promise<void> {
  let buffer = "";
  process.stdin.setEncoding("utf-8");

  process.stdin.on("data", (chunk) => {
    buffer += chunk;
  });

  process.stdin.on("end", async () => {
    const line = buffer.trim();
    if (!line) {
      process.stdout.write(
        JSON.stringify({ success: false, error: "Empty input" }) + "\n"
      );
      process.exit(0);
    }

    try {
      const req = JSON.parse(line);
      await executeOperation(req);
      process.stdout.write(JSON.stringify({ success: true }) + "\n");
      process.exit(0);
    } catch (err: any) {
      process.stdout.write(
        JSON.stringify({
          success: false,
          error: err?.message || String(err),
        }) + "\n"
      );
      process.exit(0);
    }
  });
}

main();
